### DISCUSSION TITLE
Update IP address in Tuya Local when device goes unavailable because IP changed

### DISCUSSION THREAD

**Discussion by darreni-gh1** [make-all/tuya-local#2354] [Q&A] [answered]
Hi, I'm hoping someone can help me with updating the IP address of a device in Tuya Local without deleting the device completely and re-adding it.

I have the latest version of HA Core (2024.9.3) and Tuya Local integration (2024.9.1).

I have around 30 wall switches (Tuya/Brilliant Smart Elite Glass switches 1-4 gang). Every now and then a switch will go unavailable. They were added to Tuya Local using the Smart Life app QR code and their IP's are assigned using DHCP (router doesn't support reservations :() and so the IP changes every now and then. I have tried updating the IP in the Tuya Local integration > Configure menu for the problem switch and entering the new IP address however I just get the error "Unable to connect to your device with those details. It could be an intermittent issue, or they may be incorrect." or from the debug log "Retrying after exception <class 'AttributeError'> Network Error: Unable to Connect (2/3)". I have left it for several days and tried again and still no luck.

I can't for the life of me get the switch out of this state without deleting and re-adding the device which is a pain because I have a lot of automations that need updating.

The device is still available/functional in the Smart Life app, and I can get the device details from the Tuya iot dev portal (and new device IP from my router). I am not using any other Tuya addons/integrations that would conflict.

I can ping the device on its new IP address and I can control the device with tinytuya from within the Home Assistant core container by doing the following (also works using powershell/python from my laptop). I believe I've ruled out a network issue or device id/key/version issue by doing this. I only tried this to troubleshoot after it didn't work via the integration.

docker exec -it homeassistant bash
python
import tinytuya
d = tinytuya.Device(dev_id='37138xxxxxxxxxxx9943ba', address='192.168.x.x', local_key="|-05xxxxxxxxo1g", version=3.1)
d.status()
d.turn_on()
d.turn_off()

I would have thought entering the new IP address in the Tuya Local > Configuration section for the switch would trigger a similar workflow to the above, but it just doesn't seem to work. Is this not supported, or is there any other way of updating the IP without deleting/readding the device?

I have searched the discussions and issues and read the last 10-15 or so posts including #522 which has a bunch of links for possible similar issues to what others have encountered however none of them really mention a solution for updating/changing the IP address of a device. This topic is somewhat similar however I am not referring to auto scanning but manually changing the IP: https://github.com/make-all/tuya-local/discussions/2244 

Any help would be greatly appreciated. Thanks.

**Comment by jjmonteiro** [comment]
yeah I found myself in the same scenario and had quite a hard time but managed to fix it. 💪 

I was about to ditch the whole integration as it was becoming unmaintainable, until I decided to dig into the code and try to find way around it. My idea was to pass it the device MAC address and gather the IP automatically using ARP requests until I realized it was a lot easier than that!

![image](https://github.com/user-attachments/assets/93ebb53a-2cd7-43b3-983e-355e20be6225)

The documentation says you need to have the device's IP address, turns out you don't!
The integration relies on TinyTuya, so you can use the keyword "**Auto**" in the IP address field, and it will find the device IP address automatically for you!

If it fails, simply power cycle the device and try again.
This will always keep connection with your devices in the network regardless of their IP addresses! What a bliss 🌈 

![image](https://github.com/user-attachments/assets/177bf2ac-5814-41b6-b516-1b3df6788290)

**Reply by darreni-gh1** [reply]
Appreciate the reply and suggestions. I can't use the auto option because the devices are on another subnet (sorry should have mentioned that). The tinytuya discovery/scan won't pickup the devices in this case, however tinytuya does work when specifying the IP.

My OP was more around the fact that it doesn't seem to be possible to manually change the IP address of an existing device in the integration which is strange because it appears as though you can set the device IP via the device config. @make-all is this just a know issue or limitation (or maybe I'm doing something wrong)?

I get that static/reserved IP's are preferable and I will be changing my network gear soon so I can do this, however I'll be in this situation for a few months and was hoping that I could just manually update the device IP if a device becomes unavailable because the IP changed. The devices IP's don't change too often because the DHCP lease is quite long and they usually get the same IP on renewal, but on the odd occasion the IP changes, re-adding a device is painful.

**Reply by make-all** [reply]
The intention is that IP addresses can be changed in the device options.  The only thing I can think of is that there might be some lingering dead connection there that is blocking the connection test from working after the change. If it is something like this, I'd also expect it to interfere with the Auto IP address setting - it may be the same underlying cause as the problem reported that some devices with the address set to Auto never seem to get reconnected.

**Reply by darreni-gh1** [reply]
Thanks for your response. I may take a look at the code myself if I have a chance. Appreciate the work you've done on this integration regardless, it's working very well for the most part.

**Comment by make-all** [comment]
Note that the Auto IP address setting brings its own problems, which is why the instructions no longer advertise it. Mainly slow reconnection after network failures, but in some cases the devices never seem to recover until a HA restart. Using static IP addresses or assigning DNS names to your devices is still a better option unless that absolutely is not possible with your router.

**Reply by jjmonteiro** [reply]
oh.. thanks for the heads up,
yeah, that's the prob for me, my router only allow for 5 DHCP bindings so this is still my only option 👍

**Comment by tomkerswill** [comment]
I have the same issue as the OP. My device IP changed (I've now fixed this to a new static address). And although it seems like you can type a new IP for the device, this doesn't work - it gives the same error the OP mentioned.  Disabling the device and creating a new one was the only way I could do it.

I assume it must be a bug where the IP doesn't update.   I don't really even need it to validate. If it could just update the config to the new IP, I'm pretty sure that would work.

**Reply by kbullet** [reply, marked as answer]
FYI one solution I found is if the device's ip has changed and you cannot change it on the UI and getting the unable to connect issue. Under homeassistant's .storage folder edit and change the IP address in the "**core.config_entries**" file and then restart home assistant

i.e. under /.storage/core.config_entries

Find your tuya device ip and edit it both data and options host ips e.g.

```
        "created_at": "2024-10-14T06:20:57.158790+00:00",
        "data": {
          ...
          "device_id": "xxx",
          "host": "10.0.0.90",
          ...
        },
        "domain": "tuya_local",
        "options": {
          ...
          "host": "10.0.0.90",
          ...
        },

```

hope this helps

**Reply by pilijic2** [reply]
probaste deshabilitar el dispositivo , modificar la ip y luego habilitarlo ?

**Reply by darreni-gh1** [reply]
> FYI one solution I found is if the device's ip has changed and you cannot change it on the UI and getting the unable to connect issue. Under homeassistant's .storage folder edit and change the IP address in the "**core.config_entries**" file and then restart home assistant
> 
> i.e. under /.storage/core.config_entries
> 
> Find your tuya device ip and edit it both data and options host ips e.g.
> 
> ```
>         "created_at": "2024-10-14T06:20:57.158790+00:00",
>         "data": {
>           ...
>           "device_id": "xxx",
>           "host": "10.0.0.90",
>           ...
>         },
>         "domain": "tuya_local",
>         "options": {
>           ...
>           "host": "10.0.0.90",
>           ...
>         },
> 
> ```
> 
> hope this helps 

This solution posted by @kbullet worked for me, thanks!
